README.md
# benchmark
A Python library for disaggregating time series


CHANGELOG.md
## Change Log 

### [0.3.0] - 2021-11-05

#### Maintained (operations)

- included two tests 

### [0.2.0] - 2021-11-03

#### Changed (improvement)

- allow for custom solver

### [0.1.0] - 2021-11-01

#### Added (feature)

- implemented estimator

tests/test.py
# must run pytest in current working directory

import pytest
import sys
import pathlib
import numpy as np
import pandas as pd
import statsmodels.tsa.interp

ROOT = pathlib.Path().resolve().parent
sys.path.append(str(ROOT))
from benchmark.disaggregation import Estimator

@pytest.fixture
def get_data():
    return pd.read_csv(ROOT/ "documentation/GDP.csv")

def test_imf():
    #input data
    high_frequency_data = np.array([98.2, 100.8, 102.2, 100.8, 99.0, 101.6, 102.7, 101.5, 100.5, 103.0, 103.5, 101.5])
    low_frequency_data = np.array([4000.,4161.4])
    truncation = 8

    frequency = len(high_frequency_data[:truncation]) // len(low_frequency_data)
    linear_constraint = np.ones(frequency)
    
    estimator = Estimator(linear_constraint=linear_constraint)
    estimator.fit(high_frequency_data[:truncation], low_frequency_data)
    
    interpolant = estimator.predict()
    
    imf_interpolant = np.array([969.8, 998.4, 1018.3, 1013.4, 1007.2, 1042.9, 1060.3, 1051.0, 1040.6, 1066.5, 1071.7, 1051.0])
    np.testing.assert_almost_equal(imf_interpolant[:truncation], interpolant, 1)
    
    ratio = interpolant[-1] / high_frequency_data[-1]
    extrapolant = ratio * high_frequency_data[truncation:]
    np.testing.assert_almost_equal(imf_interpolant[truncation:], extrapolant, 1)
    
def test_denton():
   #input data
    high_frequency_data = np.array([50,100,150,100] * 5)
    low_frequency_data = np.array([500,400,300,400,500])

    frequency = len(high_frequency_data) // len(low_frequency_data)
    linear_constraint = np.ones(frequency)

    estimator = Estimator(linear_constraint=linear_constraint)
    estimator.fit(high_frequency_data, low_frequency_data)

    interpolant = estimator.predict()

    denton_interpolant = np.array([64.334796,127.80616,187.82379,120.03526,56.563894,105.97568,147.50144,89.958987,40.547201,74.445963, 108.34473,76.66211,42.763347,94.14664,153.41596, 109.67405,58.290761,122.62556,190.41409,128.66959])
    np.testing.assert_almost_equal(denton_interpolant, interpolant, 1)
    
@pytest.mark.slow
def test_statsmodels(get_data):
    data = get_data

    data = data.loc[data["Country Name"].isin(["Australia", "Brazil"])].copy()
    data = data.drop(columns = ["Country Code", "Indicator Name", "Indicator Code"]).set_index("Country Name").T.copy()

    data.index = pd.to_datetime(data.index).year
    data.index.name = "Year"
    data.columns.name = "Country"

    data = data[:-1].copy()
    
    data["Brazil_Downsample"] = data["Brazil"].copy()
    data["Brazil_Downsample"][1::2] = np.NaN
    
    high_frequency_data = data["Australia"].values
    low_frequency_data = data["Brazil_Downsample"].dropna().values
    
    aud_gdp, brazil_downsample_gdp = data["Australia"].values, data["Brazil_Downsample"].dropna().values
    brazil_upsample = statsmodels.tsa.interp.dentonm(aud_gdp, brazil_downsample_gdp, freq = "other", k = 2)

    data["Brazil_Upsample"] = brazil_upsample * 2
    data[["Brazil","Brazil_Upsample"]].corr().iloc[0,1]

    estimator = Estimator()
    estimator.fit(high_frequency_data, low_frequency_data)
    
    upsampled = estimator.predict()
    np.testing.assert_almost_equal(data["Brazil_Upsample"].values, upsampled)

benchmark.py
from typing import List, Dict, Callable, Union, Optional

import numpy as np
from numpy.typing import ArrayLike

from sklearn.cross_decomposition import CCA
from sklearn.base import BaseEstimator, TransformerMixin
import sklearn.utils.validation as sklearn_validation

class Estimator(BaseEstimator, TransformerMixin):
    
    
    
    def __init__(self, linear_constraint: Optional[ArrayLike] = None, quadratic_constraint_weight: Optional[Union[float,int]] = None, solver: Optional[Callable] = None):
        self.linear_constraint = linear_constraint
        self.quadratic_constraint_weight = quadratic_constraint_weight
        self.solver = solver
         
    def fit(self, high_frequency_data: ArrayLike, low_frequency_data: ArrayLike, annotation_data: Optional[ArrayLike] = None):
        self.high_frequency_data = high_frequency_data
        self.low_frequency_data = low_frequency_data
        self.annotation_data = annotation_data
        
        self._validate()
        
        if self.solver is None:
            self.interpolant = self._solver() 
        else:
            self.interpolant = self.solver(self.high_frequency_data, self.low_frequency_data, self.annotation_data, self.linear_constraint, self.quadratic_constraint_weight) 
        
        return self

    def transform(self, X=None):
        
        model = CCA(n_components=1)
        X_c, Y_c = model.fit_transform(self.high_frequency_data.reshape(-1,self.frequency), self.low_frequency_data.reshape(-1,1))

        self.projection = model.x_rotations_.flatten()
        return np.squeeze(X_c)
    
    def predict(self, X = None):
        sklearn_validation.check_is_fitted(self, 'interpolant')
        return self.interpolant
    
    def score(self, X = None, y = None):
        sklearn_validation.check_is_fitted(self, 'interpolant')
        model = CCA(n_components=1)
        X_c, Y_c = model.fit_transform(self.interpolant.reshape(-1,self.frequency), self.low_frequency_data.reshape(-1,1))
        return np.corrcoef(np.squeeze(X_c), np.squeeze(Y_c))[0,1]
        
    def _validate(self):
        # high_frequency_data and low_frequency_data
        sklearn_validation.check_array(self.high_frequency_data, 
                                       accept_sparse = False, 
                                       dtype = "numeric", 
                                       force_all_finite=True, 
                                       ensure_2d=False, 
                                       allow_nd=False);

        self.high_frequency_data = np.asarray(self.high_frequency_data)

        if (self.high_frequency_data.ndim == 2):
            self.high_frequency_data = self.high_frequency_data.flatten()    

        if (self.solver is None) and np.any(np.isclose(self.high_frequency_data, 0)):
            raise ZeroDivisionError("high frequency data must have nonzero values for the default objective")

        sklearn_validation.check_array(self.low_frequency_data, 
                                       accept_sparse = False, 
                                       dtype = "numeric", 
                                       force_all_finite=True, 
                                       ensure_2d=False, 
                                       allow_nd=False);

        self.low_frequency_data = np.asarray(self.low_frequency_data)

        if (self.low_frequency_data.ndim != 1): 
            raise IndexError("array should be one dimensional")

        len_low = self.low_frequency_data.shape[0]
        len_high = self.high_frequency_data.shape[0]

        if (len_high % len_low != 0):
            raise IndexError("length of high frequency data must be divisible by length of low frequency data")
        else:
            self.frequency = int(len_high / len_low)

        # annotations 
        if self.annotation_data is None:
            self.annotation_data = np.full(self.high_frequency_data.shape, np.NaN)
            self.quadratic_constraint_weight = 0
        else:
            assert self.quadratic_constraint_weight is not None, "specify a value of the hyperparameter `quadradtic_constraint_weight`" 

        sklearn_validation.check_array(self.annotation_data, 
                                       accept_sparse = False, 
                                       dtype = "numeric", 
                                       force_all_finite=False, 
                                       ensure_2d=False, 
                                       allow_nd=False);

        self.annotation_data = np.asarray(self.annotation_data)

        if (self.annotation_data.ndim != 1): 
            raise IndexError("Array should be one dimensional")

        len_annotation = self.annotation_data.shape[0] 
        assert len_high == len_annotation, "length of annotation data must match length of high frequency data"

        # linear_constraint 
        if self.linear_constraint is None:
            self.linear_constraint = np.full(self.frequency, 1 / self.frequency)

        sklearn_validation.check_array(self.linear_constraint, 
                                       accept_sparse = False, 
                                       dtype = "numeric", 
                                       force_all_finite=True, 
                                       ensure_2d=False, 
                                       allow_nd=False);

        self.linear_constraint = np.asarray(self.linear_constraint)

        assert len(self.linear_constraint) == self.frequency, "length of linear constraint must equal length of high frequency data divided by length of low frequency data"

        # quadratic_constraint 
        if self.quadratic_constraint_weight is not None:
            sklearn_validation.check_scalar(self.quadratic_constraint_weight, "quadratic_constraint_weight", (float, int), min_val = 0)

    def _solver(self) -> ArrayLike:
        t = len(self.low_frequency_data)
        k = self.frequency
        T = len(self.high_frequency_data)

        constraints_matrix = np.kron(np.eye(t), self.linear_constraint.reshape(-1,1))

        inverse_high_frequency_data_matrix = np.diag(1.0 / self.high_frequency_data)
        cost_derivative_matrix = np.eye(T)

        diag_idx0, diag_idx1 = np.diag_indices(T)
        cost_derivative_matrix[diag_idx0[1:-1], diag_idx1[1:-1]] += 1
        cost_derivative_matrix[diag_idx0[:-1]+1, diag_idx1[:-1]] = -1
        cost_derivative_matrix[diag_idx0[:-1], diag_idx1[:-1]+1] = -1

        cost_matrix = 2 * np.dot(np.dot(inverse_high_frequency_data_matrix, cost_derivative_matrix), inverse_high_frequency_data_matrix)

        annotation_matrix = np.zeros(self.annotation_data.shape)
        annotation_matrix[~np.isnan(self.annotation_data)] = 1
        annotation_matrix = np.diag(annotation_matrix)
        annotation_matrix = 2 * self.quadratic_constraint_weight * annotation_matrix

        matrix = np.zeros((T+t, T+t))  
        matrix[:T,:T] = cost_matrix + annotation_matrix
        matrix[:T,T:] = constraints_matrix
        matrix[T:,:T] = constraints_matrix.T

        values = np.zeros((t+T,1)) 
        values[:T] = 2 * self.quadratic_constraint_weight * np.nan_to_num(self.annotation_data).reshape(-1,1)
        values[-t:] = self.low_frequency_data.reshape(-1,1) 

        X = np.linalg.solve(matrix, values)
        X = X[:-t] 

        return X.squeeze()

benchmark/disaggregation.py
from typing import List, Dict, Callable, Union, Optional

import numpy as np
from numpy.typing import ArrayLike

from sklearn.cross_decomposition import CCA
from sklearn.base import BaseEstimator, TransformerMixin
import sklearn.utils.validation as sklearn_validation

class Estimator(BaseEstimator, TransformerMixin):
    """ 
    An scikit-learn compatible estimator for benchmarking.
    
    Parameters
    ----------
    linear_constraint: Optional[ArrayLike] = None
        linear constraint in optimization problem. 
        default is average
        
    quadratic_constraint_weight: Optional[Union[float,int]] = None
        weight of quadratic constraint in relaxation of optimization problem
        default is 0
    
    solver: Optional[Callable] = None
        solver for custom objective in optimiation
        default is consecutive differences of ratio between up-sampled data and high-frequency data
    
    Methods
    -------
    fit: Callable 
        determine optimization problem with high frequency data and low frequency data along with annotations
    
    transform: Callable 
        projection of high frequency data through Canonical Correlation Analysis with low frequency data
    
    predict: Callable
        solution of optimization problem 
    
    score: Callable
        correlation between low frequency data and up-sampled data in Canonical Correlation Analysis
    
    Attributes
    ----------
    linear_constraint: ArrayLike
        linear constraint in optimization problem
        
    quadratic_constraint_weight: Union[float,int]]
        weight of quadratic constraint in relaxation of optimization problem
    
    solver: Callable
        solver for custom objective in optimiation
    
    high_frequency_data: ArrayLike
        high frequency dataset
        
    low_frequency_data: ArrayLike
        low frequency dataset
        
    annotations: ArrayLike
        predictions of subject matter expert

    interpolant: ArrayLike
        up-sampled dataset from solution to optimization problem

    projection: ArrayLike
        projects from `transform` method
    
    frequency: int
        Length of high_frequency_data divided by length of low_frequency_data
    
    
    Example
    -------
    >>> high_frequency_data = np.array([50,100,150,100] * 5)
    >>> low_frequency_data = np.array([500,400,300,400,500])
    >>> estimator = benchmark.Estimator()
    >>> estimator.fit(high_frequency_data, low_frequency_data)
    >>> estimator.predict()
    array([257.3392, 511.2246, 751.2952, 480.141 , 226.2556, 423.9027,590.0058, 359.8359, 162.1888, 297.7839, 433.3789, 306.6484,171.0534, 376.5866, 613.6638, 438.6962, 233.163 , 490.5022, 761.6564, 514.6784])
    
    Notes
    -----
    For more information, please consult the :ref:`User Guide <user_guide>`.
    
    See Also
    --------
    statsmodels.tsa.interp.denton
    """
    
    
    def __init__(self, linear_constraint: Optional[ArrayLike] = None, quadratic_constraint_weight: Optional[Union[float,int]] = None, solver: Optional[Callable] = None):
        self.linear_constraint = linear_constraint
        self.quadratic_constraint_weight = quadratic_constraint_weight
        self.solver = solver
         
    def fit(self, high_frequency_data: ArrayLike, low_frequency_data: ArrayLike, annotation_data: Optional[ArrayLike] = None):
        """
        fit estimator to high frequency data and low frequency data along with annotations
        
        Parameters
        ----------
        high_frequency_data: ArrayLike
            high frequency dataset
        
        low_frequency_data: ArrayLike
            low frequency dataset
        
        annotation_data: Optional[ArrayLike] = None
            predictions of subject matter expert
            
        Example
        -------
        >>> high_frequency_data = np.array([50,100,150,100] * 5)
        >>> low_frequency_data = np.array([500,400,300,400,500])
        >>> estimator = benchmark.Estimator()
        >>> estimator.fit(high_frequency_data, low_frequency_data)
        """
        self.high_frequency_data = high_frequency_data
        self.low_frequency_data = low_frequency_data
        self.annotation_data = annotation_data
        
        self._validate()
        
        if self.solver is None:
            self.interpolant = self._solver() 
        else:
            self.interpolant = self.solver(self.high_frequency_data, self.low_frequency_data, self.annotation_data, self.linear_constraint, self.quadratic_constraint_weight) 
        
        return self

    def transform(self, X=None):
        """
        project high frequency data to low frequency data through Canonical Correlation Analysis
            
        Example
        -------
        >>> high_frequency_data = np.array([50,100,150,100] * 5)
        >>> low_frequency_data = np.array([500,400,300,400,500])
        >>> estimator = benchmark.Estimator()
        >>> estimator.fit(high_frequency_data, low_frequency_data)
        >>> estimator.transform()
        
        See Also
        --------
        sklearn.cross_decomposition.CCA
        """
        
        model = CCA(n_components=1)
        X_c, Y_c = model.fit_transform(self.high_frequency_data.reshape(-1,self.frequency), self.low_frequency_data.reshape(-1,1))

        self.projection = model.x_rotations_.flatten()
        return np.squeeze(X_c)
    
    def predict(self, X = None):
        """
        return up-sampled data from solution to optimization problem
        """
        sklearn_validation.check_is_fitted(self, 'interpolant')
        return self.interpolant
    
    def score(self, X = None, y = None):
        """
        calculate correlation between up-sampled data and low frequency data through Canonical Correlation Analysis

        See Also
        --------
        sklearn.cross_decomposition.CCA
        """
        
        sklearn_validation.check_is_fitted(self, 'interpolant')
        model = CCA(n_components=1)
        X_c, Y_c = model.fit_transform(self.interpolant.reshape(-1,self.frequency), self.low_frequency_data.reshape(-1,1))
        return np.corrcoef(np.squeeze(X_c), np.squeeze(Y_c))[0,1]
        
    def _validate(self):
        # high_frequency_data and low_frequency_data
        sklearn_validation.check_array(self.high_frequency_data, 
                                       accept_sparse = False, 
                                       dtype = "numeric", 
                                       force_all_finite=True, 
                                       ensure_2d=False, 
                                       allow_nd=False);

        self.high_frequency_data = np.asarray(self.high_frequency_data)

        if (self.high_frequency_data.ndim == 2):
            self.high_frequency_data = self.high_frequency_data.flatten()    

        if (self.solver is None) and np.any(np.isclose(self.high_frequency_data, 0)):
            raise ZeroDivisionError("high frequency data must have nonzero values for the default objective")

        sklearn_validation.check_array(self.low_frequency_data, 
                                       accept_sparse = False, 
                                       dtype = "numeric", 
                                       force_all_finite=True, 
                                       ensure_2d=False, 
                                       allow_nd=False);

        self.low_frequency_data = np.asarray(self.low_frequency_data)

        if (self.low_frequency_data.ndim != 1): 
            raise IndexError("array should be one dimensional")

        len_low = self.low_frequency_data.shape[0]
        len_high = self.high_frequency_data.shape[0]

        if (len_high % len_low != 0):
            raise IndexError("length of high frequency data must be divisible by length of low frequency data")
        else:
            self.frequency = int(len_high / len_low)

        # annotations 
        if self.annotation_data is None:
            self.annotation_data = np.full(self.high_frequency_data.shape, np.NaN)
            self.quadratic_constraint_weight = 0
        else:
            assert self.quadratic_constraint_weight is not None, "specify a value of the hyperparameter `quadradtic_constraint_weight`" 

        sklearn_validation.check_array(self.annotation_data, 
                                       accept_sparse = False, 
                                       dtype = "numeric", 
                                       force_all_finite=False, 
                                       ensure_2d=False, 
                                       allow_nd=False);

        self.annotation_data = np.asarray(self.annotation_data)

        if (self.annotation_data.ndim != 1): 
            raise IndexError("Array should be one dimensional")

        len_annotation = self.annotation_data.shape[0] 
        assert len_high == len_annotation, "length of annotation data must match length of high frequency data"

        # linear_constraint 
        if self.linear_constraint is None:
            self.linear_constraint = np.full(self.frequency, 1 / self.frequency)

        sklearn_validation.check_array(self.linear_constraint, 
                                       accept_sparse = False, 
                                       dtype = "numeric", 
                                       force_all_finite=True, 
                                       ensure_2d=False, 
                                       allow_nd=False);

        self.linear_constraint = np.asarray(self.linear_constraint)

        assert len(self.linear_constraint) == self.frequency, "length of linear constraint must equal length of high frequency data divided by length of low frequency data"

        # quadratic_constraint 
        if self.quadratic_constraint_weight is not None:
            sklearn_validation.check_scalar(self.quadratic_constraint_weight, "quadratic_constraint_weight", (float, int), min_val = 0)

    def _solver(self) -> ArrayLike:
        t = len(self.low_frequency_data)
        k = self.frequency
        T = len(self.high_frequency_data)

        constraints_matrix = np.kron(np.eye(t), self.linear_constraint.reshape(-1,1))

        inverse_high_frequency_data_matrix = np.diag(1.0 / self.high_frequency_data)
        cost_derivative_matrix = np.eye(T)

        diag_idx0, diag_idx1 = np.diag_indices(T)
        cost_derivative_matrix[diag_idx0[1:-1], diag_idx1[1:-1]] += 1
        cost_derivative_matrix[diag_idx0[:-1]+1, diag_idx1[:-1]] = -1
        cost_derivative_matrix[diag_idx0[:-1], diag_idx1[:-1]+1] = -1

        cost_matrix = 2 * np.dot(np.dot(inverse_high_frequency_data_matrix, cost_derivative_matrix), inverse_high_frequency_data_matrix)

        annotation_matrix = np.zeros(self.annotation_data.shape)
        annotation_matrix[~np.isnan(self.annotation_data)] = 1
        annotation_matrix = np.diag(annotation_matrix)
        annotation_matrix = 2 * self.quadratic_constraint_weight * annotation_matrix

        matrix = np.zeros((T+t, T+t))  
        matrix[:T,:T] = cost_matrix + annotation_matrix
        matrix[:T,T:] = constraints_matrix
        matrix[T:,:T] = constraints_matrix.T

        values = np.zeros((t+T,1)) 
        values[:T] = 2 * self.quadratic_constraint_weight * np.nan_to_num(self.annotation_data).reshape(-1,1)
        values[-t:] = self.low_frequency_data.reshape(-1,1) 

        X = np.linalg.solve(matrix, values)
        X = X[:-t] 

        return X.squeeze()